### PR TITLE
Add EIP: Notification Interface

### PR DESCRIPTION
**Pull request by Oli-art**

### CODE REVIEW

File: EIPS/eip-xxxx.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
+---
+eip: xxxx
```
Comment (abcoathup) [inline EIPS/eip-xxxx.md:2]:
```suggestion
eip: 7017
```
Assigning PR number as EIP number.  Please update the filename as well.

**Comment by fonty1** [comment]
Nice proposal.. we are definitely like minded on this topic :) Streamr would be happy to help or bring exposure to this initiative if you think it can help. We are a properly decentralized pub/sub network looking to bring to life more open comms/data use cases in Web3.

Worth a look; [EthWatch](https://streamr.network/hub/projects/0x89e3a49784f0cb9f2a5b0c0dff4e8102142ef597bdcd61820c88d806ff80ca6b/overview)

**Comment by Oli-art** [comment]
> Nice proposal.. we are definitely like minded on this topic :) Streamr would be happy to help or bring exposure to this initiative if you think it can help. We are a properly decentralized pub/sub network looking to bring to life more open comms/data use cases in Web3.
> 
> Worth a look; [EthWatch](https://streamr.network/hub/projects/0x89e3a49784f0cb9f2a5b0c0dff4e8102142ef597bdcd61820c88d806ff80ca6b/overview)

Deinetely interested in some exposure. I'm developing a chrome extension that would listen for contracts implementing EIP-7017 and notify users about the notifications of their interest. I unfortunately hit a big roadblock on that one. I'm also working on a discord BOT now. For this usecases a decentralized pub/sub network is definetely useful. I'm currently using Quicknode to get the alerts to the discord BOT service.

Obiously this is all just a proof of concept and a way to test this initiative on a real-world scenario, but I would be glad to meet if you want, as I would be happy to see more projects jump on board.

**Comment by fonty1** [comment]
> I'm currently using Quicknode to get the alerts to the discord BOT service.

If its just subscribing to events then your could use the EthWatch stream - it's totally free to use and works in the browser/extension environment.

> Obiously this is all just a proof of concept and a way to test this initiative on a real-world scenario, but I would be glad to meet if you want, as I would be happy to see more projects jump on board.

Sure, feel welcome to book some time - [Calendly](https://calendly.com/matthew-fontana/one-on-one) / or reach out to me at matthew@streamr.network

**Comment by Oli-art** [comment]
> A good idea could be to implement a singleton pattern, i.e. the INotification is implemented on a fixed address and all off-chain services have to query only this address.

Yes! For wallets sending the notifications, there is really only one contract implementation of the events necesary.

In the case of contracts sending the notifications, it would't work that well, as you need to know who has sent the notification in a straight forward way, and if the contract is calling another contract, then it's more complicated to see who the real sender is. Also, if the the contract has to send many notifications, I think it's cheaper to just emit an event than to call another contract that does it for you.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
I would recommend the following changes:

1. Have a sister EIP to standardize encryption methods (so that DMMs to smart contracts can be encrypted).
2. Require encryption. This seems like a no-brainer to me.
3. Taking a book out of MMS, have the actual message parameter be a URI so the message data can be stored off-chain. If on-chain storage is desired, a `data:` URL can be used. The message parameter type would be changed to `string`, and the schema would be turned into an actual JSON schema.

Additionally, the following issues MUST be addressed:

1. The message schema must be added to the Specification. It is only mentioned in the Rationale.

**Comment by Oli-art** [comment]
> I would recommend the following changes:
> 
> 1. Have a sister EIP to standardize encryption methods (so that DMMs to smart contracts can be encrypted).
> 2. Require encryption. This seems like a no-brainer to me.
> 3. Taking a book out of MMS, have the actual message parameter be a URI so the message data can be stored off-chain. If on-chain storage is desired, a `data:` URL can be used. The message parameter type would be changed to `string`, and the schema would be turned into an actual JSON schema.
> 
> Additionally, the following issues MUST be addressed:
> 
> 1. The message schema must be added to the Specification. It is only mentioned in the Rationale.

The only point that would be a bit outside my reach is the encryption method. I found that [EIP-5630](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-5630.md) addresses this problem and seems like a good match, but I wouldn't know if it is the best or even if it's an acceptable method. Also, it hasn't been approved as an EIP.

I will be in contact with it's author to see if we can have a talk. Also, any feedback would be appreciated.

File: EIPS/eip-7017.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+The `data` field **REQUIRED** in the message events **MUST** have a JSON schema that can be written directly as a string, or passed as a URI containing the JSON. It is **RECOMENDED** to use a URI, as it is a cheaper way to send the data.
+
+It has two different specifications, one for broadcast messages and another one for direct messages. Only small differences exist between them. This is the base specifications for both:
+
+```json
+{
+"subject": "The subject of the message",
+"body": "The body of the message",
+"image_uri": "An image URI",
+"transaction_request": "A transaction request"
```
Comment (firnprotocol) [inline EIPS/eip-7017.md:134]: it looks like you are missing indentations here. would recommend trying to match lines 151–154, and also clarifying how indentation/spacing should be handled in the JSONs.

File: EIPS/eip-7017.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+The first is related to how the clients will interpret and filter the events. There can be malicious contracts that implement the notification interface but whose functions emit events where the sender is neither the message sender nor the contract. They could try to impersonate another address by changing the sender.
+
+Here is an example of a malitious implementation of a BroadcastMsg event impersonating Vitalik Buterin:
+
+```solidity
+emit BroadcastMsg (0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045, message);
+```
+
+This is why the message sender in the log's topic must allways be verified against the transaction "from" or transaction "to", depending on the notification type (wallet or contract respectively).
```
Comment (firnprotocol) [inline EIPS/eip-7017.md:275]: of course, another method would be to have a "known" trusted deployed instance of this interface, which would be hardcoded in the client-side software. that way the client code could avoid performing this check explicitly, since it trusts the contract. this seems like the route that'd be taken in practice for messaging software.

**Comment by firnprotocol** [comment]
after a conversation with Oliver, i have the following suggestions:

- remove the functions `walletDM`, `contractDM`, `walletBroadcast`, `contractBroadcast`, and only keep the _events_ `DirectMsg` and `BroadcastMsg`. it should be up to the contract to decide when, where, why and how these events get emitted; in general, the trigger need not be an explicit call by an EOA.
- remove the requirement that `from == msg.sender || from == address(this)`. in general, we want to allow arbitrary, more complex logic on the part of the contract, as far as controlling what `from` means / signifies. (for example, it could happen that a contract wants to send a message triggered by a "finalization" function, where the EOA triggering the finalization is unrelated to the sender of the message.)
- consolidate `DirectMsg` and `BroadcastMsg` into one `Message` event, where `to == 0x0000000000000000000000000000000000000000` is interpreted as as broadcast.

**Comment by firnprotocol** [comment]
> Have a sister EIP to standardize encryption methods (so that DMMs to smart contracts can be encrypted).

> Require encryption. This seems like a no-brainer to me.

@Pandapip1 [EIP-5630](https://eips.ethereum.org/EIPS/eip-5630) seems to be the appropriate standard; I will collaborating with @Oli-art to get encryption incorporated into this EIP.

**Comment by Oli-art** [comment]
I took @firnprotocol's suggestions into the last commit.

Senders are not required to be either the contract or the address executing it, although  I'm not  entierly convinced on this topic and think we would do good by hearing the opinions of people working in the wallets and dApps.

I'll ask for some meetings here and there to get more perspectives 👍

File: EIPS/eip-7017.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+- DAO governance voting anouncement
+- DEX informing an address about a certain price limit being reached, for example a stop-loss or a margin-call.
+- An NFT marketplace informing an NFT owner about an offer being made to one of it’s NFTs.
+- A metaverse informing about an important event.
+- Warning to an address about an ENS domain expiration date approaching.
+
+## Motivation
+
+With the adoption of web3 applications, an increasing necessity arises to be informed about certain events happening on-chain.
```
Comment (SamWilsn) [inline EIPS/eip-7017.md:32]: How about this wording instead:

```suggestion
As the adoption of web3 applications grows, so too does the need to be informed about certain events happening on-chain.
```

File: EIPS/eip-7017.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+Users are used to being informed, whether it be about news or updates of their favorite applications. As we are in a time of instant data feeds on social media, instant messaging apps and notifications of events that users care about, notifications are a feature in practically every application that exists in web2. They mostly come to email inboxes, SMSs, inside the applications, or in the notification inbox in the operating system.
```
Comment (SamWilsn) [inline EIPS/eip-7017.md:34]:
```suggestion
Users are accustomed to being informed, whether it be about news or updates of their favorite applications. As we are in a time of instant data feeds on social media, instant messaging apps and notifications of events that users care about, notifications are a feature in practically every application that exists in web2. They mostly come to email inboxes, SMSs, inside the applications, or in the notification inbox in the operating system.
```

File: EIPS/eip-7017.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+If they would be taken away, the engagement on these web2 applications would sink. This is not different with web3 applications: users cannot be left in the dark about what is going on in them. Not only that, for some applications, all that matters is the participation of users on certain events, like governance on a DAO.
+
+In web2, most of the user account’s are linked to an email address that is required at sign-up. This makes it easy to send notifications to specific users and requires no further complexity as an infrastructure exists to deliver a message to the user using their email address.
+
+In web3, on the other hand, there's mostly is only one inbox to send notifications to: addresses.
+Every smart contract can define its own event’s to which one can listen to, but for each of them, a change has to be done in the frontend to listen to that specific contract and event structure. This poses a problem of coordination between smart contracts and web3 applications that can notify users.
```
Comment (SamWilsn) [inline EIPS/eip-7017.md:41]:
```suggestion
Every smart contract can define its own events to which one can listen to, but for each of them, a change has to be done in the frontend to listen to that specific contract and event structure. This poses a problem of coordination between smart contracts and web3 applications that can notify users.
```

File: EIPS/eip-7017.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+- Email services
+- Discord / Telegram Bots
+
+All of these could be notifying about on-chain notifications to the user.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+An approach that is both simple, decentralized and easy to implement is to use a notifications smart contract standard that is **event-based** to be able to emit notifications to one address or broadcast them to anyone that wants to listen. Off-chain **whitelists** would record all addresses a user wants to allow receiving messages from to avoid spam. This is useful for direct messages from one address to another. Off-chain **subscription lists** would indicate which addresses they want to listen for general broadcasts. This is useful for receiving updates on a project.
```
Comment (SamWilsn) [inline EIPS/eip-7017.md:56]: This paragraph introduces no requirements or recommendations for implementation. It should be split and moved to motivation and rationale as appropriate.

File: EIPS/eip-7017.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+As a user SHALL NOT record its whitelist and subscription list on-chain, the receiver will not do any transaction.
```
Comment (SamWilsn) [inline EIPS/eip-7017.md:58]: I think this is a bit more readable:

```suggestion
Notification recipients SHALL NOT be required to record a whitelist and subscription list on-chain.
```

File: EIPS/eip-7017.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+There are security considerations for the proposal regarding impersonation threats.
+
+> It is up to the client-side application to listen only to contracts which make "sensible" choices as far as the meanings of from and to.
+
+This is related to how the clients will interpret and filter the events. There can be malicious contracts that implement the notification interface but whose functions emit events where the sender is a third party that did not consent to send the message.
+
+Here is an example of a malitious implementation of a BroadcastMsg event impersonating Vitalik Buterin:
+
+```solidity
+emit BroadcastMsg (0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045, message);
```
Comment (SamWilsn) [inline EIPS/eip-7017.md:214]: This event doesn't match one from the interface. Is that intentional?

**Comment by SamWilsn** [comment]
I am closing this pull request because we are in the process of separating EIPs and ERCs into distinct repositories. Unfortunately, as far as we are aware, GitHub does not provide any tools to ease this migration, so every pull request will need to be re-opened manually.

As this is a PR to create / modify an ERC, I will kindly ask you to redirect this to the new repository at [ethereum/ERCs](https://github.com/ethereum/ERCs). **We have prepared [a guide](https://github.com/ethereum/ERCs/issues/8) to help with the process.**

If there is relevant history here, please link to this PR from the new pull request.

On behalf of the EIP Editors, I apologize for this inconvenience.

**Comment by wakqasahmed** [comment]
@SamWilsn @Oli-art has this EIP moved to the next stage? how close it is to be incorporated as a standard?
